Review 4 updates

copejon opened the pull request:
Major Changes:

- Dropped COSIRegistration
- BucketContent API should snapshot BucketClass
- `objectBucketName` expanded meaning to `bucketIdentifier`

copejon committed:
bucketContents capture class fields

copejon committed:
replace objectBucketName with bucketIdentifier

copejon committed:
reworked case table, static requires secret and bucketID b/c secret no longer has ep

copejon committed:
convert tabs to spaces

copejon committed:
updated toc

copejon committed:
removed COSIRegistration definition

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -160,7 +158,7 @@ Sidecar start up will follow these steps:
 1. Controller creates a `BucketContent` object with its `BucketClassName` set to the name of its `BucketClass` and a `finalizer`.
 1. Sidecar detects the new `BucketContent` object and gets the associated `BucketClass`.
-1. Sidecar calls the GrantBucketAccess() rpc, passing the `bucketClass.objectBucketName` and the `bucketClass.parameters` and is returned a bucket endpoint and credentials.
+1. Sidecar calls the GrantBucketAccess() rpc, passing the `bucketClass.bucketIdentifier` and the `bucketClass.parameters` and is returned a bucket endpoint and credentials.
```
should this be an interface to handle different identifiers for different object stores? Can we do that in an API?

copejon replied:
It's possible in go to marshal null interfaces, but the behavior is a little unpredictable (something that goes in as an int comes out as a float64 for example).  Couple that with the fact that we'd be passing it through a gRPC interface to something possibly written in another lang and I don't want to think about it.

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,27 +255,44 @@ Metadata:
 1. `name`: Generated in the pattern of `<BUCKET-CLASS-NAME>'-'<RANDOM-SUFFIX>`. 
 1. `labels`: COSI controller adds the label to its managed resources for easy CLI GET ops.  Value is the driver name returned by GetDriverInfo() rpc. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
 1. `finalizers`: COSI controller adds the finalizer to defer Bucket deletion until backend deletion ops succeed.
+1. `provisioner`: The name of the drive as defined in the BucketClass.  Used by sidecars to filter BucketContents.
```
maybe add `Copied from the BucketClass` to these new fields?

copejon replied:
👍

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -299,7 +314,7 @@ provisioner: [1]
 accessMode: {"ro", "wo", "rw"} [3]
 releasePolicy: {"Delete", "Retain"} [4]
-objectBucketName: [5]
+bucketIdentifier: [5]
```
Consider adding to `provisioner` below:
`Provisioner names follow the DNS naming conventions, eg. "ceph-rgw.cosi.redhat.com"`

copejon replied:
Agree, but since it's appears in a few places (and likely will appear more once we start defining the gRPC interface), I'll call it out explicitly in one place

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,27 +255,44 @@ Metadata:
+    - `name` : the Bucket’s name
+    - `namespace`: the Bucket’s namespace
+    - `uuid`: the Bucket’s API server generated UUID
+    - `resourceVersion`: the Bucket’s 
```
I know this field is commonly added here, but how is it useful if we have the uuid? Is it for supporting >1 COSI version simultaneously? Is it for zero-downtime upgrades?

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,27 +255,44 @@ Metadata:
 1. `secretRef`: the name and namespace of the source secret in the `Bucket`'s namespace.  This `secret` may be generated by the driver or created by an admin.
-1. `objectBucketName`: the name of the actual bucket in the object store.
-1. `bucketAttributes`: stateful data relevant to the managing of the bucket but potentially inappropriate user knowledge (e.g. user's IAM role name).
+1. `bucketIdentifier`: a unique set of identifying information 
```
do have to define this set? Can you provide an example?

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,27 +255,44 @@ Metadata:
-1. `objectBucketName`: the name of the actual bucket in the object store.
-1. `bucketAttributes`: stateful data relevant to the managing of the bucket but potentially inappropriate user knowledge (e.g. user's IAM role name).
+1. `bucketIdentifier`: a unique set of identifying information 
+1. `message`: a human readable description detailing the reason for the current `phase`
```
how does this differ from `reason`? And why do we no longer need the `bucketAttributes` statefull data?

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,27 +255,44 @@ Metadata:
 1. `secretRef`: the name and namespace of the source secret in the `Bucket`'s namespace.  This `secret` may be generated by the driver or created by an admin.
-1. `objectBucketName`: the name of the actual bucket in the object store.
-1. `bucketAttributes`: stateful data relevant to the managing of the bucket but potentially inappropriate user knowledge (e.g. user's IAM role name).
+1. `bucketIdentifier`: a unique set of identifying information 
```
Design considerations:
The id of a bucket may be more complex than a simple string name or url endpoint. SDKs often expect separate args for the name, port, region and whatever else they embed into a bucket url. An app should not have to decompose a full bucket endpoint (url) so that it can call SDK access functions. A full endpoint seems to be only useful for debugging / curl calls, outside of the app.

COSI also should not have to dig into an opaque type and figure out where the bucket name is, etc. in order to compose a secret or file content to be ingested by the app pod. Thus, in order to generate endpoint related fields and a secret, a map of these values makes sense.

Drivers also need to save some state info (eg. IAM user) so that they can de-provision a bucket that they created, and the BucketContent is a reasonable location for that. All "state" data should be ignored by COSI other than storing it in the BucketContent and passing it to the driver. 

Here are BucketContent field suggestions:
 +  `objectBucketName: <string>` simply names the bucket and nothing else. It is likely redundant with one of the key-values provided in `bucketEndpointAttributes` below.
+   `bucketEndpointAttributes: <map[string]string>` contains all of the values needed to fully define the bucket's endpoint, eg BUCKET=, REGION=, PORT=, etc. These keys and values are provisioner-specific and are added to the secret.
 +  `provisionerState: <base64 string blob>` contains provisioner-specific state values. Eg the IAM user may be needed to de-provision a bucket and thus would be contained in this field. COSI never looks at this value but will received it from the driver, store it in the BucketContent, and pass it to the driver for _DeleteBucket_ and _RevokeBucketAccess_ rpc calls.

copejon replied:
The `bucketEndpointAttributes` would also have to be returned to the driver.  That's analogous to the `bucketIdentifier` Saad had mentioned.

copejon committed:
bucketAttributes removed - drive can retrieve data from store

bucketEndpoint removed - cannot guarantee value here

bucketIdentify now expected to be sent to workload.  up to app to decode

copejon committed:
remove bucketIdentifier from Bucket, bucketContent is referenced by Bucket

copejon committed:
bucketClass formatting fixes

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -227,10 +212,10 @@ metadata:
   bucketPrefix: [3]
   bucketClassName: [4]
-  secretName: [5]
+  secretName: [6]
```
number 5?

copejon replied:
dang numbers.

copejon committed:
fixed ordered list formatting

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,36 +242,54 @@ Metadata:
 status:
-  bucketAttributes: <map[string]string> [8]
-  phase: [9]
+  message: [14]
```
Number 13 here and 14 below

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,36 +242,54 @@ Metadata:
+1. `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield and Static cases, *Retain* is mandated.
+    - _Delete_:  the bucket and its contents are destroyed
+    - _Retain_:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed
+1. `accessModes`: Declares the level of access given to credentials provisioned through this class.     If empty, drivers may set defaults.
```
Intentionally plural? I don't see the accessModes defined here.
nit formatting: number of spaces between "." and "If empty"

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,36 +242,54 @@ Metadata:
+    - _Delete_:  the bucket and its contents are destroyed
+    - _Retain_:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed
+1. `accessModes`: Declares the level of access given to credentials provisioned through this class.     If empty, drivers may set defaults.
+1. `supportedProtocols`:  An array of protocols the associated object store supports (e.g. swift, s3, gcs, etc.). *Only* serves a descriptive purpose and is not verified.
```
do we want it to be a formal array (- yaml notation)? Or is a string w/ comma separators better?

copejon replied:
i would think an array, rather than having to parse a string

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,36 +242,54 @@ Metadata:
-1. `secretRef`: the name and namespace of the source secret in the `Bucket`'s namespace.  This `secret` may be generated by the driver or created by an admin.
-1. `objectBucketName`: the name of the actual bucket in the object store.
-1. `bucketAttributes`: stateful data relevant to the managing of the bucket but potentially inappropriate user knowledge (e.g. user's IAM role name).
+    - `name` : the Bucket’s name
```
This is basically an objectRef, which (I think) is being deprecated. Is this still what we want here?

copejon replied:
Yes, definitely.  This structure, and it's go counterpart, are well defined.  ObjectRefs were deprecated because they were loose in what they could describe, which is what made them very hard to validate and document.

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -299,37 +302,26 @@ provisioner: [1]
 
-1. `provisioner`: (Optional) The name of the driver. If supplied the driver container and sidecar container are expected to be deployed. If omitted the `secretRef` is required for static provisioning.
-1. `supportedProtocols`: (Optional) An array of protocols the associated object store supports (e.g. swift, s3, gcs, etc.). *Only* serves a descriptive purpose and is not verified.
+1. `provisioner`: The name of the driver. If supplied the driver container and sidecar container are expected to be deployed. If omitted the `secretRef` is required for static provisioning.
```
Is this required now? What is its value in the static brownfield case?

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -299,37 +302,26 @@ provisioner: [1]
 
-1. `provisioner`: (Optional) The name of the driver. If supplied the driver container and sidecar container are expected to be deployed. If omitted the `secretRef` is required for static provisioning.
-1. `supportedProtocols`: (Optional) An array of protocols the associated object store supports (e.g. swift, s3, gcs, etc.). *Only* serves a descriptive purpose and is not verified.
+1. `provisioner`: The name of the driver. If supplied the driver container and sidecar container are expected to be deployed. If omitted the `secretRef` is required for static provisioning.
```
Why are the two "(Optional)" tags deleted here but remain for `accessModes`?

copejon replied:
Protocol allow users to be certain of what they're getting.  Without the field, we have to assume that the class is just named in some way that would clue in users to the supported clients.

Provisioner is mandatory as well - the only time it would allowable to be undefined would be for default bucket classes, which I thought we agreed didn't make sense.  The reason they don't is because we decided we weren't going to provide any matching (by protocol, accesssMode, etc) to classes.

jeffvance commented on keps/sig-storage/20191125-bucket-provisioning.md:
```diff
@@ -257,36 +242,54 @@ Metadata:
-    - _Released_: the `Bucket` has been deleted, signalling that the `BucketContent` is ready for garbage collection.
-    - _Failed_: error and all retries have been exhausted.
-    - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.
+     - Bound_: the controller finished processing the request and bound the `Bucket` and `BucketContent`_
```
leading "_" needed for italics dropped...

copejon committed:
accessModes are singular because a client operate as more than one at a time
